listener.go

```diff
@@ -105,46 +106,37 @@ func (l *listener) handleConn(conn net.Conn) {
 
 func (l *listener) Accept() (net.Conn, error) {
 	select {
-	case conn, ok := <-l.nextConn:
-		if !ok {
-			return nil, ErrClosed
-		}
+	case <-l.chClosed:
+		return nil, ErrClosed
+	case conn := <-l.nextConn:
 		return conn, nil
-	case err, ok := <-l.nextErr:
-		if !ok {
-			return nil, ErrClosed
-		}
+	case err := <-l.nextErr:
 		return nil, err
 	}
 }
 
-func (l *listener) Close() error {
-	l.mx.Lock()
-	defer l.mx.Unlock()
-	if l.closed {
-		return nil
-	}
-	for _, session := range l.sessions {
-		closeErr := session.Close()
-		if closeErr != nil {
-			log.Errorf("Error closing session: %v", closeErr)
+func (l *listener) Close() (err error) {
+	l.closeOnce.Do(func() {
+		close(l.chClosed)
+		for _, session := range l.sessions {
+			closeErr := session.Close()
+			if closeErr != nil {
+				log.Errorf("Error closing session: %v", closeErr)
+			}
 		}
-	}
-	l.closed = true
-	err := l.Listener.Close()
-	// Drain nextConn and nextErr
-drain:
-	for {
-		select {
-		case conn := <-l.nextConn:
-			conn.Close()
-		case <-l.nextErr:
-		default:
-			break drain
+		err = l.Listener.Close()
+		// Drain nextConn and nextErr
+		for {
+			select {
+			case conn := <-l.nextConn:
+				conn.Close()
+			case <-l.nextErr:
```

Comment:
The bug was here. After `l.nextErr` being closed, this branch was always selected, hence an infinite loop.

Reply:
It might be one of the culprit when Lantern occupies too much CPU sometimes (when the proxies list is reset).

Reply:
Not really. It's listener code which never runs at client.